### ISSUE TITLE
Minimal viable trinity-beacon node

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/trinity#240]
## Abstract

This issue tries to document the current state of the `trintiy-beacon` command and lays out steps to turn it into an actual minimal viable beacon node. This was sparked by @ChihChengLiang starting this [document](https://notes.ethereum.org/znnddsMNTLikDHVrkVosSQ?both) which means it is also trying to address questions outlined there.

### Background: How does the beacon node implementation fit into the code base?

In general, the relation between the beacon node and the existing Ethereum 1 client is very loose at this point.

1. The `trinity-beacon` command is an entirely different command from `trinity` and invokes its own bootstrapping routine at `trinity/main_beacon.py`. Bootstrapping code does share some common utility functions.

2. The long term vision for Trinity is to be mostly a composition of plugins with many essential parts just being written as builtin, pre-loaded core plugins (while other use defined plugins can be loaded at runtime)

We can see this already working out with the collection of current plugins.

https://github.com/ethereum/trinity/blob/7436d9c3fdef5fbeef5d7a1123c8ed2ecd16e033/trinity/plugins/registry.py#L49-L68

The plugin infrastructure also takes care of spawning isolated processes for each plugin that derives from `BaseIsolatedPlugin` and enables inter-process communication through a central `EventBus`.

3. Trinity provides a `TrinityConfig` that should be agnostic of anything that is specific to the Ethereum1 or beacon node. These specific configs need to go into the `BeaconAppConfig` which will be available in the context of the beacon node (similar to the `Eth1AppConfig` being available in the context of the Ethereum 1 node).

4. Currently, the beacon node uses all base plugins (**not** the Ethereum 1 specific plugins) which includes things like the *Peer Discovery* as well as the *JSON-RPC API*.

### Does the beacon node need a `Node` class?

Probably not. The `Node` / `FullNode` / `LightNode` classes are currently used in the Ethereum 1 node to run what is currently known as the `networking` process. This process does currently host the `PeerPool`, the server that answers request from other peers as well as the syncing code.

**We are actively migrating away from that architecture** towards an architecture where the responsibilities are further broken down into different processes that are assembled through plugins.

Ideally, we'll soon have something that looks like this:

1. Peer Pool Plugin (own process) -> manages peers, relays incoming messages on the `EventBus`

2. Sync Plugin (own process) -> runs the sync, interacts with the `PeerPool` via the `EventBus` (probably hidden a convenience API) ([see comment](https://github.com/ethereum/trinity/pull/213))

3. Chainserver Plugin (own process) -> serve other clients,  interacts with the `PeerPool` via the `EventBus`

When that happens, we won't have a hardcoded `networking` process anymore and instead have three processes all bootstrapped through plugins just like the discovery and JSON-RPC are already today.

Most likely, this does also mean there won't be a `Node` class anymore as these plugins will use their more focused abstractions.

### Sooo?

For the beacon node, there are two possible ways from here:

1. Accelerate the re-architecture process by jumping on #213 to get to that new architecture sooner (I'm on it, but I encourage others to also experiment in the same direction if they want to)

2. Otherwise, just create a new plugin `BeaconNodePlugin` that just bootstraps a `PeerPool`, server, syncer in more or less the same way as the `networking` process currently does. This would be temporary code that would then later be broken down into smaller plugins as soon as we have a `PeerPool` that supports being run in its dedicated process and integrates natively with the `EventBus`. ([see comment](https://github.com/ethereum/trinity/pull/155#issuecomment-456811014))

### Where do we stand with the database process?

The beacon node does have its database process bootstrapped but we still need to handle the initialization of the database when its empty. 

### Do we need a `database_mode`?

The `database_mode` is used in the Ethereum 1 node to differentiate between `light` and `full` nodes that use an incompatible database. The beacon node should be free to handle database access in its own way however it is needed. This is why the `database_mode` is a property on the `Eth1AppConfig` and not on the `TrinityConfig`.

I believe in later phase, the beacon node would at least manage one database for the beacon chain and `n` databases for the shard. I currently do not know if there needs to be something like a different database for light nodes?

**Comment by cburgdorf** [comment]
This is a quick brain dump. Happy to discuss and fill in the holes!

/cc @ChihChengLiang @hwwhww @jannikluhn @pipermerriam @djrtwo

**Comment by ChihChengLiang** [comment]
IIUC, to use the new architecture, the best timing to implement Sync Plugin and Chainserver Plugin is when Peer Pool Plugin is completed?

**Comment by cburgdorf** [comment]
@ChihChengLiang that is correct but I wouldn't want you being blocked on that. So, I would propose that you just move forward with the Peer Pool, Chain Server and Sync all running in one process similar to the current architecture of the Ethereum 1 node.

To do that, you can either:

1. Implement a `Node` class similar to the one currently being used for Ethereum 1 and then just kick that off in a `BaseIsolatedPlugin`.

2. Just forget the `Node` class and throw together a `BaseIsolatedPlugin` that bootstraps the `PeerPool`, Server and Syncer in an ad hoc way (comparable to how it is assembled under tests)

Then, later, when we have a `PeerPool` that can run in its own isolated plugin with an EventBus-based API consumable from other processes, we can migrate to use that and to move chain server and sync into distinct plugins as well.

**Comment by jannikluhn** [comment]
Thank you very much for this detailed write up. It cleared up a couple of things for me.

> I currently do not know if there needs to be something like a different database for light nodes?

I'd say we should ignore light nodes for a long time.

**Comment by cburgdorf** [comment]
I believe the current beacon node implementation is already past this and we can close this issue.
